src/utils/array.ts

```diff
@@ -0,0 +1,139 @@
+type ItemCallback<T, U> = (item: T, index: number, array: T[]) => U
```

Comment:
did you copy this file from another project? When you do, please give credit and be careful about licenses.

Reply:
All code written by me, in this file as well as any files I've modified, are my own original work.

The array functions stem from these utility functions I wrote: 
https://github.com/snuffyDev/tint/tree/master/src/collections/array

And the Object `deepClone` function is also based on my own deep merge utility:
https://github.com/snuffyDev/tint/blob/master/src/collections/objects.ts#L67